nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py

```diff
@@ -655,3 +654,122 @@ def reset_tools(self):
+
+    The original paper refer to: https://arxiv.org/abs/1804.03294.
+    """
+
+    def __init__(self, model: Module, config_list: List[Dict], trainer: Callable[[Module, Optimizer, Callable], None],
+                 optimizer: Optimizer, criterion: Callable[[Tensor, Tensor], Tensor], iterations: int, training_epochs: int):
+        """
+        Parameters
+        ----------
+        model
+            Model to be pruned.
+        config_list
+            Supported keys:
+                - sparsity : This is to specify the sparsity for each layer in this config to be compressed.
+                - sparsity_per_layer : Equals to sparsity.
+                - rho : Penalty parameters in ADMM algorithm.
+                - op_types : Operation types to prune.
+                - op_names : Operation names to prune.
+                - exclude : Set True then the layers setting by op_types and op_names will be excluded from pruning.
+        trainer
+            A callable function used to train model or just inference. Take model, optimizer, criterion as input.
+            The model will be trained or inferenced `training_epochs` epochs.
+
+            Example::
+
+                def trainer(model: Module, optimizer: Optimizer, criterion: Callable[[Tensor, Tensor], Tensor]):
+                    training = model.training
+                    model.train(mode=True)
+                    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+                    for batch_idx, (data, target) in enumerate(train_loader):
+                        data, target = data.to(device), target.to(device)
+                        optimizer.zero_grad()
+                        output = model(data)
+                        loss = criterion(output, target)
+                        loss.backward()
+                        # If you don't want to update the model, you can skip `optimizer.step()`, and set train mode False.
+                        optimizer.step()
+                    model.train(mode=training)
+        optimizer
+            The optimizer instance used in trainer. Note that this optimizer might be patched during collect data,
+            so do not use this optimizer in other places.
+        criterion
+            The criterion function used in trainer. Take model output and target value as input, and return the loss.
+        iterations
+            The total iteration number in admm pruning algorithm.
+        training_epochs
+            The epoch number for training model in each iteration.
+        """
+        self.trainer = trainer
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.iterations = iterations
+        self.training_epochs = training_epochs
+        super().__init__(model, config_list)
+
+        self.Z = {name: wrapper.module.weight.data.clone().detach() for name, wrapper in self.get_modules_wrapper().items()}
```

Comment:
please use a clear name rather than `Z` and `U`.

Reply:
`U` can be named as `scaled_dual_variable`, but I have no idea to name `Z`. The author rewrites the origin problem to 2-block optimization, and `Z` can be seen as another solution of `weight`. I think using `Z` is because in ADMM, the second optimization goal is usually denoted as `z`.
![image](https://user-images.githubusercontent.com/33053116/133359431-e9aadcd8-dfb1-4c67-a0f7-7bb5c3c00624.png)

And in compression, the problem is
![image](https://user-images.githubusercontent.com/33053116/133359556-ab7eedbd-d5a9-4fee-832b-ac4c4d2ca157.png)

Reply:
Got it.